tensorflow_addons/optimizers/lamb_optimizer.py

```diff
@@ -0,0 +1,220 @@
+
+        # m_t = beta1 * m + (1 - beta1) * g_t
+        m = self.get_slot(var, 'm')
+        m_scaled_g_values = grad * coefficients['one_minus_beta_1_t']
+        m_t = m.assign(
+            m * coefficients['beta_1_t'], use_locking=self._use_locking)
+        with tf.control_dependencies([m_t]):
+            m_t = self._resource_scatter_add(m, indices, m_scaled_g_values)
+
+        # v_t = beta2 * v + (1 - beta2) * (g_t * g_t)
+        v = self.get_slot(var, 'v')
+        v_scaled_g_values = (grad * grad) * coefficients['one_minus_beta_2_t']
+        v_t = v.assign(
+            v * coefficients['beta_2_t'], use_locking=self._use_locking)
+        with tf.control_dependencies([v_t]):
+            v_t = self._resource_scatter_add(v, indices, v_scaled_g_values)
+
+        m_t_hat = m_t / (1. - coefficients['beta_1_power'])
+        v_t_hat = v_t / (1. - coefficients['beta_2_power'])
+
+        v_sqrt = tf.sqrt(v_t_hat)
+        update = m_t_hat / (v_sqrt + coefficients['epsilon'])
+
+        var_name = self._get_variable_name(var.name)
+        if self._do_use_weight_decay(var_name):
+            update += coefficients['weight_decay_rate'] * var
+
+        ratio = 1.0
+        if self._do_layer_adaptation(var_name):
+            w_norm = tf.norm(var, ord=2)
+            g_norm = tf.norm(update, ord=2)
+            ratio = tf.where(
+                tf.greater(w_norm, 0),
+                tf.where(tf.greater(g_norm, 0), (w_norm / g_norm), 1.0), 1.0)
+
+        var_update = var.assign_sub(
+            ratio * coefficients['lr_t'] * update,
+            use_locking=self._use_locking)
+        return tf.group(*[var_update, m_t, v_t])
+
+    def get_config(self):
+        config = super(LAMBOptimizer, self).get_config()
+        config.update({
+            'learning_rate':
+            self._serialize_hyperparameter('learning_rate'),
+            'weight_decay_rate':
+            self._serialize_hyperparameter('weight_decay_rate'),
+            'decay':
+            self._serialize_hyperparameter('decay'),
+            'beta_1':
+            self._serialize_hyperparameter('beta_1'),
+            'beta_2':
+            self._serialize_hyperparameter('beta_2'),
+            'epsilon':
+            self.epsilon,
+        })
+        return config
+
+    def _do_use_weight_decay(self, param_name):
+        """Whether to use L2 weight decay for `param_name`."""
+        if self.exclude_from_weight_decay:
+            for r in self.exclude_from_weight_decay:
+                if re.search(r, param_name) is not None:
+                    return False
+        return True
+
+    def _do_layer_adaptation(self, param_name):
+        """Whether to do layer-wise learning rate adaptation for
+        `param_name`."""
+        if self.exclude_from_layer_adaptation:
+            for r in self.exclude_from_layer_adaptation:
+                if re.search(r, param_name) is not None:
+                    return False
```

Comment:
This may be more clear once `exclude_from_layer_adaptation` and `exclude_from_weight_decay` are documented, but in general regex is rarely used in TF-Core; and almost never used in Keras subclasses

Reply:
https://github.com/tensorflow/tensorflow/search?q=%22import+re%22&unscoped_q=%22import+re%22